### PR TITLE
Add support for conditionally task in pipeline visualization

### PR DESCRIPTION
**Pull request by karthikjeeyar**
**Fixes**: 

https://issues.redhat.com/browse/ODC-5152

**Analysis / Root cause**: 

As a user, I want to get visual cues in the pipeline diagram for conditional tasks, so that I can understand when a task conditional and when in execution to see if it was Skipped or Passed.

**Solution Description**: 

Added when decorator for the condtionals tasks and on hovering the decorator the state of the condition is displayed in the tooltip.

**Screen shots / Gifs for design review**: 

Pipeline page:
image

PipelineRuns page:
image

Tooltip:
image

Mp4:

https://user-images.githubusercontent.com/9964343/113874756-be357480-97d3-11eb-8363-7084bfc391cf.mp4

**Unit test coverage report**: 

 WhenExpressionDecorator:

    ✓ should render diamond shape when expression decorator (8ms)
    ✓ should not append a line after the diamond shape when appendLine prop is not passed (2ms)
    ✓ should append a line after the diamond shape when appendLine prop is passed (2ms)
    ✓ should not render tooltip when enableTooltip prop is set to false (12ms)
    ✓ should render tooltip when enableTooltip prop is set to true (9ms)
    ✓ should contain the succeeded tooltip content if the task status is succeeded (7ms)
    ✓ should contain the skipped tooltip content if the task status is skipped (6ms)
    ✓ should contain the default tooltip content for other task status (14ms)

Test Suites: 1 passed, 1 total
Tests:       8 passed, 8 total

  hasWhenExpression:
```
    ✓ expect to return false if the pipeline does not contain when expression (1ms)
    ✓ expect to return true if the regular tasks in the pipeline contains when expression
    ✓ expect to return true if the finally tasks in the pipeline contains when expression
```
    
**Test setup:**

Install openshift pipelines 1.4 operator to test the when expression in finally tasks.

**Browser conformance**: 

- [x] Chrome
- [x] Firefox
- [ ] Safari
- [ ] Edge

/kind feature
cc: @andrewballantyne @bgliwa01

### CODE REVIEW

**Review by bgliwa01 — APPROVED** [review]
this looks great, Karthik!!

**Comment by andrewballantyne** [comment]
Not sure this is a desired outcome for regular visualizations:
Screen Shot 2021-04-07 at 2 59 15 PM

cc @bgliwa01

File: frontend/packages/pipelines-plugin/src/components/pipelines/pipeline-topology/WhenExpressionDecorator.tsx [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+  stroke = lightBorderColor.value,
+  appendLine = false,
+  status,
+}) => {
+  const { t } = useTranslation();
+  const rotation = 45; // 45deg
+  const diamondHeight =
+    Math.round(width * Math.sin(rotation)) + Math.round(height * Math.cos(rotation));
+  const diamondNode = (
+    <>
```
Comment (andrewballantyne) [inline frontend/packages/pipelines-plugin/src/components/pipelines/pipeline-topology/WhenExpressionDecorator.tsx:34]: don't need fragments on solo nodes.

File: frontend/packages/pipelines-plugin/src/components/pipelines/pipeline-topology/WhenExpressionDecorator.tsx [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+            y1={diamondHeight / 2}
+            x2={width + height}
+            y2={diamondHeight / 2}
+            stroke={lightBorderColor.value}
+          />
+        )}
+      </g>
+    </>
+  );
+  const getTooltipContent = () => {
```
Comment (andrewballantyne) [inline frontend/packages/pipelines-plugin/src/components/pipelines/pipeline-topology/WhenExpressionDecorator.tsx:49]: Another function, almost IIFE here, any reason why not static?

**Review by andrewballantyne — CHANGES_REQUESTED** [review]

File: frontend/packages/pipelines-plugin/src/components/pipelines/detail-page-tabs/pipeline-details/PipelineVisualizationTask.tsx [hunk 3]
```diff
@@ -151,6 +163,12 @@ const TaskComponent: React.FC<TaskProps> = ({
     status?.reason !== runStatus.Pending &&
     status?.reason !== runStatus.Cancelled &&
     !!path;
+  const hasWhenExpression = pipelineTask?.when?.length > 0;
+  const hasRunAfter = pipelineTask?.runAfter?.length > 0;
+  const getTaskColor = (): string =>
+    status
+      ? getRunStatusColor(status.reason, t).pftoken.value
+      : getRunStatusColor(runStatus.Cancelled, t).pftoken.value;
```
[the remark this answers is not in the document]
Comment (karthikjeeyar) [reply frontend/packages/pipelines-plugin/src/components/pipelines/detail-page-tabs/pipeline-details/PipelineVisualizationTask.tsx:171]: Initially I was passing the status in the function,  while refactoring I realized the status is available in global scope, so removed the argument from the function. I will update both the places that uses functions.

File: frontend/packages/pipelines-plugin/src/components/pipelines/detail-page-tabs/pipeline-details/PipelineVisualizationTask.tsx [hunk 4]
```diff
@@ -215,22 +231,42 @@ const TaskComponent: React.FC<TaskProps> = ({
       )}
     </g>
   );
+
+  const taskColor = showStatusState
+    ? getTaskColor()
+    : !isFinallyTask
+    ? greyBackgroundColor.value
+    : lightBackgroundColor.value;
```
[the remark this answers is not in the document]
Comment (karthikjeeyar) [reply frontend/packages/pipelines-plugin/src/components/pipelines/detail-page-tabs/pipeline-details/PipelineVisualizationTask.tsx:239]: I am filling the background to avoid the underlying line here.

[the remark this answers is not in the document]
Comment (andrewballantyne) [reply frontend/packages/pipelines-plugin/src/components/pipelines/detail-page-tabs/pipeline-details/PipelineVisualizationTask.tsx:239]: Gotcha, thanks!

File: frontend/packages/pipelines-plugin/src/components/pipelines/pipeline-topology/WhenExpressionDecorator.tsx [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+  switch (status) {
+    case runStatus.Succeeded:
+      tooltipContent = <div>{t('pipelines-plugin~When expression was met')}</div>;
+      break;
+    case runStatus.Skipped:
+      tooltipContent = <div>{t('pipelines-plugin~When expression was not met')}</div>;
+      break;
+    default:
+      tooltipContent = <div>{t('pipelines-plugin~When expression')}</div>;
+      break;
```
Comment (andrewballantyne) [inline frontend/packages/pipelines-plugin/src/components/pipelines/pipeline-topology/WhenExpressionDecorator.tsx:57]: surprised this didn't pop up as a lint error... but `break` is not needed in default.

**Review by andrewballantyne — CHANGES_REQUESTED** [review]
Okay, can you do one final change... let's just do a quick rundown of the graph looking for when. If it's not there, we can reset the distance between nodes. It's just easier I think that way. We need to really rework our dagre layout to support proper balances as UX wants, but until then, this will have to do.

**Comment by karthikjeeyar** [comment]
Fixed tasks without when expression to fallback to existing layout options.

image

File: frontend/packages/pipelines-plugin/src/components/pipelines/pipeline-topology/utils.ts [hunk 6]
```diff
@@ -304,12 +310,19 @@ export const getTopologyNodesEdges = (
   return { nodes, edges };
 };
 
+export const hasWhenExpression = (pipeline: PipelineKind): boolean => {
+  return [...(pipeline?.spec?.tasks ?? []), ...(pipeline?.spec?.finally ?? [])].some(
```
Comment (andrewballantyne) [inline frontend/packages/pipelines-plugin/src/components/pipelines/pipeline-topology/utils.ts:314]: fwiw, `||` is probably more applicable here since `tasks` and `finally` are only ever arrays. If you could have `''` or `0` or `false` as valid values, it would make sense to do `??` but if you're either "not defined" (null / undefined) or an array, it's easier to just do `||` (imo).

Entirely a nit, arguably just a preference, and definitely not important.

**Review by andrewballantyne — APPROVED** [review]
We'll have to readdress the layout issues for sure... [ODC-3129](https://issues.redhat.com/browse/ODC-3129) covers some of this desire and may end up morphing to handle all the various use-cases.

File: frontend/packages/pipelines-plugin/src/components/pipelines/pipeline-topology/WhenExpressionDecorator.scss [hunk 7]
```diff
@@ -0,0 +1,3 @@
+.opp-diamond {
```
Comment (abhinandan13jan) [inline frontend/packages/pipelines-plugin/src/components/pipelines/pipeline-topology/WhenExpressionDecorator.scss:1]: Should we not use `opp-filename-diamond`?

File: frontend/packages/pipelines-plugin/src/components/pipelines/pipeline-topology/__tests__/WhenExpressionDecorator.spec.tsx [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+  return {
+    ...reactI18next,
+    useTranslation: () => ({ t: (key) => key }),
+  };
+});
+
+type WhenExpressionDecoratorProps = React.ComponentProps<typeof WhenExpressionDecorator>;
+
+describe('WhenExpressionDecorator', () => {
+  let wrapper: ShallowWrapper<WhenExpressionDecoratorProps>;
```
Comment (abhinandan13jan) [inline frontend/packages/pipelines-plugin/src/components/pipelines/pipeline-topology/__tests__/WhenExpressionDecorator.spec.tsx:19]: nit: `whenExpressionDecoratorWrapper`?

Comment (karthikjeeyar) [reply frontend/packages/pipelines-plugin/src/components/pipelines/pipeline-topology/__tests__/WhenExpressionDecorator.spec.tsx:19]: TBH,  `whenExpressionDecoratorWrapper` name is too long and wrapper is used to in many spots and i think we dont need to be specific about the wrapper as we are having only one component in this test.

**Comment by abhinandan13jan** [comment]
@karthikjeeyar Maybe one issue we need to address from backend. Not related to the PR but an edge case when a Task got removed post execution, one task shows pending after pipeline completed execution cc: @andrewballantyne 
Screenshot from 2021-04-08 16-56-36

The PR works fine and looks good
/LGTM

**Comment by karthikjeeyar** [comment]
Information about removed task should be available in the Pipelinerun yaml so that we can reflect it in the UI.
